[Inductor][HOP][print]Add inductor and IR class for print

fxdawnn opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #170880
* __->__ #167680
## Problem
In the hop print setup, we weren't able to get inductor running with the straightforward with_effect setup. 

## Solutions
Thus, we use ``fallbackkernel`` as a media to re-route the cpp and python hop print for its most efficient ways by (1) python: removing the hop in inductor so as to directly called python``print()`` without dispather and not to cause further graph break; (2) use ``cout`` print in c++.

With the with_effect inductor handling, hop print could easily be adopted. 

- cpp codegen for inductor with cout style print [Removed and saved for next PR]

- python codegen for inductor to remove hop (so as not to go through dispather) and assert builtin.print to avoid graph break.
<img width="724" height="412" alt="Screenshot 2025-12-19 at 11 28 18 AM" src="https://github.com/user-attachments/assets/b9a27887-bf51-4b14-a48e-ec9cc83326b2" />

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @mlazos @chenyang78

fxdawnn committed:
[Inductor][HOP][print]Add inductor and IR class for print

[ghstack-poisoned]

angelayi commented on test/higher_order_ops/test_print.py:
```diff
@@ -254,5 +254,17 @@ def fn(a, b):
         )
 
 
+class TestHopPrintInductor(TestCase):
```
please remove this too.

angelayi commented on test/higher_order_ops/test_print.py:
```diff
@@ -254,5 +254,17 @@ def fn(a, b):
+        inp = [torch.randn(3, 3), torch.ones(3, 3)]
+        ref_out = fn(*inp)
+        # Validate the hop print can reduce the graph break in dynamo tracing
+        out = torch.compile(backend="inductor")(fn)(*inp)
```
please check that the print happens

fxdawnn committed:
Update on "[Inductor][HOP][print]Add inductor and IR class for print"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

fxdawnn committed:
Update on "[Inductor][HOP][print]Add inductor and IR class for print"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

fxdawnn committed:
Update on "[Inductor][HOP][print]Add inductor and IR class for print"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

fxdawnn committed:
Update on "[Inductor][HOP][print]Add inductor and IR class for print"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

fxdawnn committed:
Update on "[Inductor][HOP][print]Add inductor and IR class for print"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

fxdawnn committed:
Update on "[Inductor][HOP][print]Add inductor and IR class for print"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

fxdawnn committed:
Update on "[Inductor][HOP][print]Add inductor and IR class for print"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

fxdawnn committed:
Update on "[Inductor][HOP][print]Add inductor and IR class for print"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

fxdawnn committed:
Update on "[Inductor][HOP][print]Add inductor and IR class for print"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

fxdawnn commented on torch/_inductor/ir.py:
```diff
@@ -8732,6 +8732,87 @@ def codegen(self, wrapper: PythonWrapperCodegen) -> None:
+        )
+
+        # Write the print call
+        wrapper.writeline(f"{kernel_name}({repr(self.format_str)}, {kwargs_code})")
```
yes, I've added that. I'll come back to this later to figure out why this print is not happening in test tho haha

fxdawnn committed:
Update on "[Inductor][HOP][print]Add inductor and IR class for print"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

fxdawnn committed:
Update on "[Inductor][HOP][print]Add inductor and IR class for print"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

fxdawnn committed:
Update on "[Inductor][HOP][print]Add inductor and IR class for print"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

fxdawnn committed:
Update on "[Inductor][HOP][print]Add inductor and IR class for print"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

fxdawnn committed:
Update on "[Inductor][HOP][print]Add inductor and IR class for print"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

fxdawnn committed:
Update on "[Inductor][HOP][print]Add inductor and IR class for print"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

fxdawnn committed:
Update on "[Inductor][HOP][print]Add inductor and IR class for print"



With the with_effect inductor handling, hop print could easily be adopted.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

Lucaskabela reviewed (changes requested):
Generally looks good but there seem to be many comments left in that can be removed

fxdawnn commented on torch/_inductor/ir.py:
```diff
@@ -8738,6 +8738,70 @@ def codegen(self, wrapper: PythonWrapperCodegen) -> None:
+            kwargs=kwargs_for_op,
+            python_kernel_name="torch.ops.higher_order.print",
+        )
+        # self.format_str = format_str
```
So these are commented out because we were unsure of the with_effect inductor impl. But we gained more clarity after last week. I'll update this later!

fxdawnn commented on torch/_inductor/ir.py:
```diff
@@ -8738,6 +8738,70 @@ def codegen(self, wrapper: PythonWrapperCodegen) -> None:
+        return [NoneAsConstantBuffer()]
+
+    def codegen(self, wrapper: PythonWrapperCodegen) -> None:
+        # wrapper.writeline(f'print("{self.format_str}".format(**self.print_kwargs))')
```
It will be finalized soon!

fxdawnn committed:
Update on "[Inductor][HOP][print]Add inductor and IR class for print"



With the with_effect inductor handling, hop print could easily be adopted.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

fxdawnn committed:
Update on "[Inductor][HOP][print]Add inductor and IR class for print"



With the with_effect inductor handling, hop print could easily be adopted.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

fxdawnn committed:
Update on "[Inductor][HOP][print]Add inductor and IR class for print"



With the with_effect inductor handling, hop print could easily be adopted.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

fxdawnn committed:
Update on "[Inductor][HOP][print]Add inductor and IR class for print"



With the with_effect inductor handling, hop print could easily be adopted.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

fxdawnn committed:
Update on "[Inductor][HOP][print]Add inductor and IR class for print"



With the with_effect inductor handling, hop print could easily be adopted.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

fxdawnn committed:
Update on "[Inductor][HOP][print]Add inductor and IR class for print"



With the with_effect inductor handling, hop print could easily be adopted.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

fxdawnn committed:
Update on "[Inductor][HOP][print]Add inductor and IR class for print"



With the with_effect inductor handling, hop print could easily be adopted.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

fxdawnn committed:
Update on "[Inductor][HOP][print]Add inductor and IR class for print"



With the with_effect inductor handling, hop print could easily be adopted.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

fxdawnn committed:
Update on "[Inductor][HOP][print]Add inductor and IR class for print"



With the with_effect inductor handling, hop print could easily be adopted.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

Lucaskabela commented on torch/_inductor/codegen/cpp_wrapper_cpu.py:
```diff
@@ -2349,13 +2352,15 @@ def extract_output_name(
+            else:
+                # For non-CallTorchBind HOPs (like print), we don't need schema for C++ wrapper
+                # since they're handled specially (e.g., printf for print)
+                return_schema = []
```
since print is the only one that should use this path currently, could we explicitly assert here that the op is  `torch.ops.higher_order.print`?

fxdawnn replied:
That's a great point. Initially (before print), this part seems to only support ``torch._higher_order_ops.torchbind.CallTorchBind``. I'll specify the error handling if it's neither torchbind nor print.

fxdawnn committed:
Update on "[Inductor][HOP][print]Add inductor and IR class for print"



With the with_effect inductor handling, hop print could easily be adopted.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

fxdawnn committed:
Update on "[Inductor][HOP][print]Add inductor and IR class for print"



With the with_effect inductor handling, hop print could easily be adopted.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

angelayi commented on test/higher_order_ops/test_print.py:
```diff
@@ -243,6 +244,99 @@ def f(x):
+        code uses builtins.print directly rather than calling torch.ops.higher_order.print,
+        which is more efficient and avoids unnecessary overhead.
+        """
+        from torch._inductor.utils import run_and_get_code
```
import at top level?

angelayi commented on test/higher_order_ops/test_print.py:
```diff
@@ -243,6 +244,99 @@ def f(x):
+        C++ wrapper uses std::cout which writes to file descriptor 1 (stdout).
+        We need to redirect the actual file descriptor to capture the output.
+        """
+        import os
```
same here?

angelayi commented on test/higher_order_ops/test_print.py:
```diff
@@ -243,6 +244,99 @@ def f(x):
+        inputs = (torch.randn(2, 3),)
+
+        # Enable C++ wrapper and capture stdout at the file descriptor level
+        with config.patch({"cpp_wrapper": True}):
```
can we also check with `run_and_get_code` here?

fxdawnn replied:
I'll add one!

angelayi commented on torch/_inductor/codegen/cpp_wrapper_cpu.py:
```diff
@@ -2371,6 +2380,49 @@ def extract_output_name(
             assert isinstance(output_name, list), type(output_name)
             output_args = output_name
 
+        # Special handling for torch.ops.higher_order.print
```
can we create a dictionary like "custom_extern_kernel_codegen" that maps op to codegen implementation for cpp and python? this way we can keep all the code in one place. And then here we would just check if the op is in in the dictionary then we call that custom implementation

cc @BoyuanFeng

angelayi commented on torch/_inductor/lowering.py:
```diff
@@ -7296,6 +7296,20 @@ def cond(
+    result = ir.FallbackKernel.create(hop_print, format_str, **kwargs)
+    # print returns None, but FallbackKernel.create may return None or []
+    # Ensure we return a list for with_effects to unpack
+    return result if result is not None else []
```
is this still needed?

fxdawnn replied:
not needed. updated.

angelayi commented on torch/_inductor/ir.py:
```diff
@@ -8244,6 +8254,7 @@ def __init__(
         nontensor_args: Sequence[Any],
         unflatten_args: Callable[..., Any],
         *,
+        kwargs: Optional[dict[str, Any]] = None,
```
it's not clear to me why youre running into a ComplexView

fxdawnn committed:
Update on "[Inductor][HOP][print]Add inductor and IR class for print"



With the with_effect inductor handling, hop print could easily be adopted.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

fxdawnn committed:
Update on "[Inductor][HOP][print]Add inductor and IR class for print"



With the with_effect inductor handling, hop print could easily be adopted. 

- cpp codegen for inductor with cout style print
<img width="1224" height="566" alt="Screenshot 2025-12-19 at 11 22 14 AM" src="https://github.com/user-attachments/assets/f401efb5-575f-4485-aabc-37deb6e80ab1" />

- python codegen for inductor to remove hop (so as not to go through dispather) and assert builtin.print to avoid graph break.
- 
[
<img width="724" height="412" alt="Screenshot 2025-12-19 at 11 28 18 AM" src="https://github.com/user-attachments/assets/b9a27887-bf51-4b14-a48e-ec9cc83326b2" />
]

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

Lucaskabela commented on torch/_inductor/ir.py:
```diff
@@ -8249,6 +8259,7 @@ def __init__(
         nontensor_args: Sequence[Any],
         unflatten_args: Callable[..., Any],
         *,
```
nit: `dict[str, Any] | None` for python 3.10 typing

Lucaskabela reviewed (commented):
Will defer to Angela on the special handling logic, but rest looks good

fxdawnn committed:
Update on "[Inductor][HOP][print]Add inductor and IR class for print"

## Problem
In the hop print setup, we weren't able to get inductor running with the straightforward with_effect setup. 

## Solutions
Thus, we use ``fallbackkernel`` as a media to re-route the cpp and python hop print for its most efficient ways by (1) python: removing the hop in inductor so as to directly called ``builtin.print`` without dispather and not to cause further graph break; (2) use ``cout`` print in c++.

With the with_effect inductor handling, hop print could easily be adopted. 

- cpp codegen for inductor with cout style print
<img width="1224" height="566" alt="Screenshot 2025-12-19 at 11 22 14 AM" src="https://github.com/user-attachments/assets/f401efb5-575f-4485-aabc-37deb6e80ab1" />

- python codegen for inductor to remove hop (so as not to go through dispather) and assert builtin.print to avoid graph break.
<img width="724" height="412" alt="Screenshot 2025-12-19 at 11 28 18 AM" src="https://github.com/user-attachments/assets/b9a27887-bf51-4b14-a48e-ec9cc83326b2" />

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

fxdawnn committed:
Update on "[Inductor][HOP][print]Add inductor and IR class for print"

## Problem
In the hop print setup, we weren't able to get inductor running with the straightforward with_effect setup. 

## Solutions
Thus, we use ``fallbackkernel`` as a media to re-route the cpp and python hop print for its most efficient ways by (1) python: removing the hop in inductor so as to directly called ``builtin.print`` without dispather and not to cause further graph break; (2) use ``cout`` print in c++.

With the with_effect inductor handling, hop print could easily be adopted. 

- cpp codegen for inductor with cout style print [Removed and saved for next PR]

- python codegen for inductor to remove hop (so as not to go through dispather) and assert builtin.print to avoid graph break.
<img width="724" height="412" alt="Screenshot 2025-12-19 at 11 28 18 AM" src="https://github.com/user-attachments/assets/b9a27887-bf51-4b14-a48e-ec9cc83326b2" />

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

fxdawnn committed:
Update on "[Inductor][HOP][print]Add inductor and IR class for print"

## Problem
In the hop print setup, we weren't able to get inductor running with the straightforward with_effect setup. 

## Solutions
Thus, we use ``fallbackkernel`` as a media to re-route the cpp and python hop print for its most efficient ways by (1) python: removing the hop in inductor so as to directly called ``builtin.print`` without dispather and not to cause further graph break; (2) use ``cout`` print in c++.

With the with_effect inductor handling, hop print could easily be adopted. 

- cpp codegen for inductor with cout style print [Removed and saved for next PR]

- python codegen for inductor to remove hop (so as not to go through dispather) and assert builtin.print to avoid graph break.
<img width="724" height="412" alt="Screenshot 2025-12-19 at 11 28 18 AM" src="https://github.com/user-attachments/assets/b9a27887-bf51-4b14-a48e-ec9cc83326b2" />

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

fxdawnn committed:
Update on "[Inductor][HOP][print]Add inductor and IR class for print"

## Problem
In the hop print setup, we weren't able to get inductor running with the straightforward with_effect setup. 

## Solutions
Thus, we use ``fallbackkernel`` as a media to re-route the cpp and python hop print for its most efficient ways by (1) python: removing the hop in inductor so as to directly called ``builtin.print`` without dispather and not to cause further graph break; (2) use ``cout`` print in c++.

With the with_effect inductor handling, hop print could easily be adopted. 

- cpp codegen for inductor with cout style print [Removed and saved for next PR]

- python codegen for inductor to remove hop (so as not to go through dispather) and assert builtin.print to avoid graph break.
<img width="724" height="412" alt="Screenshot 2025-12-19 at 11 28 18 AM" src="https://github.com/user-attachments/assets/b9a27887-bf51-4b14-a48e-ec9cc83326b2" />

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

angelayi commented on torch/_inductor/codegen/custom_extern_kernel_codegen.py:
```diff
@@ -0,0 +1,109 @@
+    dict[str, Callable[[ir.FallbackKernel, Callable[[str], None]], None]],
+] = {
+    "torch.ops.higher_order.print": {
+        "python": generate_print_python,
```
instead of a dictionary can you make a dataclass or named tuple? so we can directly access it, like "custom_kernel_codegen[op].python"

fxdawnn committed:
Update on "[Inductor][HOP][print]Add inductor and IR class for print"

## Problem
In the hop print setup, we weren't able to get inductor running with the straightforward with_effect setup. 

## Solutions
Thus, we use ``fallbackkernel`` as a media to re-route the cpp and python hop print for its most efficient ways by (1) python: removing the hop in inductor so as to directly called ``builtin.print`` without dispather and not to cause further graph break; (2) use ``cout`` print in c++.

With the with_effect inductor handling, hop print could easily be adopted. 

- cpp codegen for inductor with cout style print [Removed and saved for next PR]

- python codegen for inductor to remove hop (so as not to go through dispather) and assert builtin.print to avoid graph break.
<img width="724" height="412" alt="Screenshot 2025-12-19 at 11 28 18 AM" src="https://github.com/user-attachments/assets/b9a27887-bf51-4b14-a48e-ec9cc83326b2" />

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

fxdawnn committed:
Update on "[Inductor][HOP][print]Add inductor and IR class for print"

## Problem
In the hop print setup, we weren't able to get inductor running with the straightforward with_effect setup. 

## Solutions
Thus, we use ``fallbackkernel`` as a media to re-route the cpp and python hop print for its most efficient ways by (1) python: removing the hop in inductor so as to directly called ``builtin.print`` without dispather and not to cause further graph break; (2) use ``cout`` print in c++.

With the with_effect inductor handling, hop print could easily be adopted. 

- cpp codegen for inductor with cout style print [Removed and saved for next PR]

- python codegen for inductor to remove hop (so as not to go through dispather) and assert builtin.print to avoid graph break.
<img width="724" height="412" alt="Screenshot 2025-12-19 at 11 28 18 AM" src="https://github.com/user-attachments/assets/b9a27887-bf51-4b14-a48e-ec9cc83326b2" />

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

fxdawnn committed:
Update on "[Inductor][HOP][print]Add inductor and IR class for print"

## Problem
In the hop print setup, we weren't able to get inductor running with the straightforward with_effect setup. 

## Solutions
Thus, we use ``fallbackkernel`` as a media to re-route the cpp and python hop print for its most efficient ways by (1) python: removing the hop in inductor so as to directly called ``builtin.print`` without dispather and not to cause further graph break; (2) use ``cout`` print in c++.

With the with_effect inductor handling, hop print could easily be adopted. 

- cpp codegen for inductor with cout style print [Removed and saved for next PR]

- python codegen for inductor to remove hop (so as not to go through dispather) and assert builtin.print to avoid graph break.
<img width="724" height="412" alt="Screenshot 2025-12-19 at 11 28 18 AM" src="https://github.com/user-attachments/assets/b9a27887-bf51-4b14-a48e-ec9cc83326b2" />

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

angelayi commented on torch/_inductor/codegen/custom_extern_kernel_codegen.py:
```diff
@@ -0,0 +1,104 @@
+    codegen_kwargs = node.codegen_kwargs()
+
+    # First arg is the format string
+    format_str = codegen_args[0] if codegen_args else "''"
```
if there are no codegen args shouldnt that be an error?

fxdawnn committed:
Update on "[Inductor][HOP][print]Add inductor and IR class for print"

## Problem
In the hop print setup, we weren't able to get inductor running with the straightforward with_effect setup. 

## Solutions
Thus, we use ``fallbackkernel`` as a media to re-route the cpp and python hop print for its most efficient ways by (1) python: removing the hop in inductor so as to directly called ``builtin.print`` without dispather and not to cause further graph break; (2) use ``cout`` print in c++.

With the with_effect inductor handling, hop print could easily be adopted. 

- cpp codegen for inductor with cout style print [Removed and saved for next PR]

- python codegen for inductor to remove hop (so as not to go through dispather) and assert builtin.print to avoid graph break.
<img width="724" height="412" alt="Screenshot 2025-12-19 at 11 28 18 AM" src="https://github.com/user-attachments/assets/b9a27887-bf51-4b14-a48e-ec9cc83326b2" />

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

fxdawnn committed:
Update on "[Inductor][HOP][print]Add inductor and IR class for print"

## Problem
In the hop print setup, we weren't able to get inductor running with the straightforward with_effect setup. 

## Solutions
Thus, we use ``fallbackkernel`` as a media to re-route the cpp and python hop print for its most efficient ways by (1) python: removing the hop in inductor so as to directly called ``builtin.print`` without dispather and not to cause further graph break; (2) use ``cout`` print in c++.

With the with_effect inductor handling, hop print could easily be adopted. 

- cpp codegen for inductor with cout style print [Removed and saved for next PR]

- python codegen for inductor to remove hop (so as not to go through dispather) and assert builtin.print to avoid graph break.
<img width="724" height="412" alt="Screenshot 2025-12-19 at 11 28 18 AM" src="https://github.com/user-attachments/assets/b9a27887-bf51-4b14-a48e-ec9cc83326b2" />

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

fxdawnn committed:
Update on "[Inductor][HOP][print]Add inductor and IR class for print"

## Problem
In the hop print setup, we weren't able to get inductor running with the straightforward with_effect setup. 

## Solutions
Thus, we use ``fallbackkernel`` as a media to re-route the cpp and python hop print for its most efficient ways by (1) python: removing the hop in inductor so as to directly called ``builtin.print`` without dispather and not to cause further graph break; (2) use ``cout`` print in c++.

With the with_effect inductor handling, hop print could easily be adopted. 

- cpp codegen for inductor with cout style print [Removed and saved for next PR]

- python codegen for inductor to remove hop (so as not to go through dispather) and assert builtin.print to avoid graph break.
<img width="724" height="412" alt="Screenshot 2025-12-19 at 11 28 18 AM" src="https://github.com/user-attachments/assets/b9a27887-bf51-4b14-a48e-ec9cc83326b2" />

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

angelayi commented on torch/_inductor/codegen/custom_extern_kernel_codegen.py:
```diff
@@ -0,0 +1,123 @@
+    return result
+
+
+def has_custom_codegen(op_name: str) -> bool:
```
this is not used?

fxdawnn replied:
good catch!

fxdawnn committed:
Update on "[Inductor][HOP][print]Add inductor and IR class for print"

## Problem
In the hop print setup, we weren't able to get inductor running with the straightforward with_effect setup. 

## Solutions
Thus, we use ``fallbackkernel`` as a media to re-route the cpp and python hop print for its most efficient ways by (1) python: removing the hop in inductor so as to directly called ``builtin.print`` without dispather and not to cause further graph break; (2) use ``cout`` print in c++.

With the with_effect inductor handling, hop print could easily be adopted. 

- cpp codegen for inductor with cout style print [Removed and saved for next PR]

- python codegen for inductor to remove hop (so as not to go through dispather) and assert builtin.print to avoid graph break.
<img width="724" height="412" alt="Screenshot 2025-12-19 at 11 28 18 AM" src="https://github.com/user-attachments/assets/b9a27887-bf51-4b14-a48e-ec9cc83326b2" />

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

fxdawnn commented on torch/_inductor/codegen/custom_extern_kernel_codegen.py:
```diff
@@ -0,0 +1,123 @@
+}
+
+
+def get_custom_codegen(
```
good catch!

angelayi commented on torch/_inductor/codegen/custom_extern_kernel_codegen.py:
```diff
@@ -0,0 +1,96 @@
+CUSTOM_EXTERN_KERNEL_CODEGEN: dict[str, CustomCodegen] = {
+    "torch.ops.higher_order.print": CustomCodegen(
+        python=generate_print_python,
+        # cpp=generate_print_cpp,  # Add C++ implementation when needed
```
delete?

fxdawnn committed:
Update on "[Inductor][HOP][print]Add inductor and IR class for print"

## Problem
In the hop print setup, we weren't able to get inductor running with the straightforward with_effect setup. 

## Solutions
Thus, we use ``fallbackkernel`` as a media to re-route the cpp and python hop print for its most efficient ways by (1) python: removing the hop in inductor so as to directly called python``print()`` without dispather and not to cause further graph break; (2) use ``cout`` print in c++.

With the with_effect inductor handling, hop print could easily be adopted. 

- cpp codegen for inductor with cout style print [Removed and saved for next PR]

- python codegen for inductor to remove hop (so as not to go through dispather) and assert builtin.print to avoid graph break.
<img width="724" height="412" alt="Screenshot 2025-12-19 at 11 28 18 AM" src="https://github.com/user-attachments/assets/b9a27887-bf51-4b14-a48e-ec9cc83326b2" />

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo chenyang78

[ghstack-poisoned]

fxdawnn commented:
@pytorchbot merge
